sway/input/cursor.c

```diff
@@ -544,7 +544,19 @@ static void handle_tablet_tool_position(struct sway_cursor *cursor,
 	node_at_coords(seat, cursor->cursor->x, cursor->cursor->y, &surface, &sx, &sy);
 	struct sway_tablet_tool *sway_tool = tool->data;
 
-	if (surface && wlr_surface_accepts_tablet_v2(tablet->tablet_v2, surface)) {
+	// The logic for whether we should send a tablet event or an emulated pointer
+	// event is tricky. It comes down to:
+	// * If we began a drag on a non-tablet surface (simulated_tool_tip_down),
+	//   then we should continue sending emulated pointer events regardless of
+	//   whether the surface currently under us accepts tablet or not.
+	// * Otherwise, if we are over a surface that accepts tablet, then we should
+	//   send tablet events.
+	// * If we began a drag over a tablet surface, we should continue sending
+	//   tablet events until the drag is released, even if we are now over a
+	//   non-tablet surface.
```

Comment:
Can't we send events in the `tablet_tool_tip` handlers instead?

Reply:
Hmm, I'm not sure I follow... do you mean within the seatop handlers for `tablet_tool_tip`? We do send the actual motion events in the seatop handler for `tablet_tool_motion`, but deciding whether to emulate input as pointer events is a task that seems best left to `cursor.c`, so that all the special-casing can happen in one place as opposed to over different seatop implementations.

That leaves the seatops receiving only pointer and tablet input, without having to worry about whether it's _emulated_ pointer input they're receiving, or "real" input.

Reply:
I agree that emulating pointer events is something that should be done in `cursor.c`. However calling functions that directly send events like `wlr_tablet_v2_tablet_tool_notify_proximity_out` is a little fragile here. The seatop may or may not forward events to `wlr_seat`, in case it doesn't, we shouldn't notify proximity out.

But I see we don't have `sway_seatop_impl.tablet_tool_proximity_out` yet, so my point is moot anyway. Also looking at this PR now, the diff doesn't touch the code near `wlr_tablet_v2_tablet_tool_notify_proximity_out` (though maybe that has changed with the rebase? can't remember).

tl;dr LGTM!